sql/src/main/java/io/crate/action/job/ContextPreparer.java

```diff
@@ -156,16 +305,44 @@ public ExecutionSubContext visitCountPhase(CountPhase phase, PreparerContext con
         public ExecutionSubContext visitMergePhase(final MergePhase phase, final PreparerContext context) {
             RamAccountingContext ramAccountingContext = RamAccountingContext.forExecutionPhase(circuitBreaker, phase);
 
-            RowReceiver downstream = getDownstream(context, phase,
-                    Paging.getWeightedPageSize(Paging.PAGE_SIZE, 1.0d / phase.executionNodes().size()));
+            boolean upstreamOnSameNode = context.getPhaseHasSameNodeUpstream(phase.executionPhaseId(), (byte) 0);
+
+            int pageSize = Paging.getWeightedPageSize(Paging.PAGE_SIZE, 1.0d / phase.executionNodes().size());
+            RowReceiver rowReceiver = context.getRowReceiver(phase, pageSize);
+            if (rowReceiver == null) {
+                context.executionPhasesToProcess.add(phase);
+                return null;
+            }
+
+            if (upstreamOnSameNode) {
+                if (!phase.projections().isEmpty()) {
+                    ProjectorChainContext projectorChainContext = new ProjectorChainContext(phase.name(), rowReceiver);
```

Comment:
could the FlatProjectorChain be created inside the ProjectorChainContext? that way you wouldn't need the projectorChain setter